FINERACT-1744 - Idempotency support

b0c1 opened the pull request:
## Description

FINERACT-1744 - Idempotency support
- [x] Idempotency write move to filters
- [x] Idempotency integration test with success and failure
- [x] Added response status code to database
- [x] SynchronousCommandProcessingService not generate the command, just marked to the filter

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [x] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [x] Create/update unit or integration tests for verifying the changes made.

- [x] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [x] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [x] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

galovics commented on fineract-provider/src/main/java/org/apache/fineract/commands/service/IdempotencyKeyResolver.java:
```diff
@@ -0,0 +1,52 @@
+
+    private final FineractProperties fineractProperties;
+
+    public IdempotencyKeyResolver(IdempotencyKeyGenerator idempotencyKeyGenerator, FineractProperties fineractProperties) {
```
Could be replaced by @RequiredArgsConstructor.

galovics commented on fineract-provider/src/main/java/org/apache/fineract/commands/service/SynchronousCommandProcessingService.java:
```diff
@@ -64,83 +67,70 @@ public class SynchronousCommandProcessingService implements CommandProcessingSer
     private final IdempotencyKeyGenerator idempotencyKeyGenerator;
-    private final FineractProperties fineractProperties;
+    private final CommandSourceService commandSourceService;
+    private final Gson gson = new Gson();
```
Replace this direct instantiation with `org.apache.fineract.infrastructure.core.serialization.GoogleGsonSerializerHelper#createSimpleGson`

galovics commented on fineract-provider/src/main/java/org/apache/fineract/commands/service/SynchronousCommandProcessingService.java:
```diff
@@ -149,15 +139,39 @@ public CommandProcessingResult executeCommand(final CommandWrapper wrapper, fina
+    }
+
+    private void setIdempotencyKeyStoreFlag(boolean flag) {
+        RequestContextHolder.currentRequestAttributes().setAttribute("storeIdempotencyKey", flag, RequestAttributes.SCOPE_REQUEST);
```
Magic string (the attribute key), please extract it to a constant.

galovics commented on fineract-provider/src/main/java/org/apache/fineract/commands/service/SynchronousCommandProcessingService.java:
```diff
@@ -64,83 +67,70 @@ public class SynchronousCommandProcessingService implements CommandProcessingSer
 
         final boolean rollbackTransaction = configurationDomainService.isMakerCheckerEnabledForTask(wrapper.taskPermissionName());
+        String idempotencyKey = idempotencyKeyResolver.resolve(wrapper);
+        checkExistingCommand(wrapper, idempotencyKey);
```
Well, I understand this method (checkExistingCommand) but I don't really like how this is structured in terms of execution flow. If I just look at this (executeCommand) method, it's not very clear that this checkExistingCommand will actually break the execution flow by throwing exceptions.

Would be nicer to restructure this a bit so it's more cleaner.

galovics commented on fineract-provider/src/main/java/org/apache/fineract/commands/service/SynchronousCommandProcessingService.java:
```diff
@@ -64,83 +67,70 @@ public class SynchronousCommandProcessingService implements CommandProcessingSer
+        if (savedCommandSource.getId() == null) {
+            throw new IllegalStateException("Command source not saved");
+        }
+        RequestContextHolder.currentRequestAttributes().setAttribute("commandSourceId", savedCommandSource.getId(),
```
`commandSourceId` is a magic string. Constant pls.

galovics replied:
Also, would be nice to extract this line to a separate method so it can be more descriptive.

b0c1 committed:
FINERACT-1744 - Idempotency support
- [x] Idempotency write move to filters
- [x] Idempotency integration test with success and failure
- [x] Added response status code to database
- [x] SynchronousCommandProcessingService not generate the command, just marked to the filter
- [x] IdempotencyStoreFilter store the command request if available.
- [x] Removed JpaExceptionHandler.java

galovics commented on fineract-provider/src/main/java/org/apache/fineract/commands/handler/NewCommandSourceHandler.java:
```diff
@@ -20,8 +20,12 @@
 
 public interface NewCommandSourceHandler {
 
+    @Transactional(propagation = Propagation.REQUIRES_NEW, isolation = Isolation.REPEATABLE_READ)
```
I don't remember. What if the handler implementor also uses @Transactional? Does it override this behavior or not?

galovics commented on fineract-provider/src/main/java/org/apache/fineract/infrastructure/core/exceptionmapper/DuplicateCommandExceptionMapper.java:
```diff
@@ -0,0 +1,61 @@
+    @Override
+    public Response toResponse(final DuplicateCommandException exception) {
+        log.debug("Duplicate request: {}", exception.getMessage());
+        if (exception instanceof CommandProcessedException) {
```
Not a fan of this. Can we do separate ExceptionMappers for each type of Exception type?

galovics commented on integration-tests/src/test/java/org/apache/fineract/integrationtests/common/IdempotencyHelper.java:
```diff
@@ -0,0 +1,99 @@
+        return new Gson().fromJson(json, new TypeToken<JobBusinessStepConfigData>() {}.getType());
+    }
+
+    // private static ApiParameterError configuredApiParameterErrorFromJsonString(final String json) {
```
Comment.

galovics commented on fineract-provider/src/main/java/org/apache/fineract/infrastructure/core/exception/AbstractIdempotentCommandException.java:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    @Override
+    public String getMessage() {
```
ToString hasnt been updated. Do you mind using a @ToString on the class?

galovics commented on fineract-provider/src/main/java/org/apache/fineract/infrastructure/core/exceptionmapper/DuplicateCommandExceptionMapper.java:
```diff
@@ -0,0 +1,41 @@
+
+    @Override
+    public Response toResponse(final DuplicateCommandException exception) {
+        return Response.status(Status.INTERNAL_SERVER_ERROR).entity(serverSideError("500", "")).type(MediaType.APPLICATION_JSON).build();
```
Since we're returning an HTTP 500 here, I think we should log the exception itself at least so we can see the stacktrace. Don't you think?

b0c1 replied:
Class removed
